### ISSUE TITLE
list ament_python as setup_requires to ensure installation

### ISSUE THREAD

**Issue by nuclearsandwich** [ament/ament_python#3]
Following up on [this comment](https://github.com/ros2/rosdistro/issues/7#issuecomment-311463970) from @tfoote.

Requiring the presence of a non-standard python module in setup.py means that we'll always require end users to manually install a dependency before they can set up and use an ament_python package.

For from-source builds, some of the capabilities handled by ament_python are handled by the ament build tool instead, but thus far the tool is not, nor is it intended to be, required to build ROS 2 projects (or other projects using ament but lacking examples I'm not inclined to consider them at the moment.)

The two capabilities provided by ament_python currently are
1. the placement of data files into the ament package index
1. fudging of the install-scripts location to install executables into a libexec directory rather than the provided or default bin directory.

Both of these are value-adds that improve interoperability with other ROS 2 components but are otherwise not required for the functionality provided by the python packages using this.

I'm not familiar enough with the Python ecosystem to say whether or not there is a way to provide the capabilities that ament_python and ament_tools currently do without requiring something to be present before setuptools resolves dependencies. But supposing that there is no such alternative and also supposing that we do not want to make mandatory the use of ament_tools, I think we can continue to use the ament_python package to provide additional install-time capabilities as long as we document for package creators how to gracefully degrade when ament_python is not available.

Something along the lines of

```python
try:
    from ament_python import install_scripts_to_libexec
    install_scripts_to_libexec(package_name)
except ImportError:
    print("WARNING: without ament_python certain functionality will be disabled")
```

I am not proposing to guarantee that expected ament / ROS functionality and interoperability will still work when ament_python is not present, if we could do that, we would not need ament_python at all. Just that if setup.py is invoked when ament_python is unavailable, it still successfully installs a python package.

Doing this would resolve the outstanding issue on the ROS 2 buildfarm, because ament_python's absence would no longer crash setup.py while running `setup.py clean` during sourcedeb production.

**Comment by clalancette** [comment]
That feels kind of kludgy, and somewhat difficult for any random python package to conform to (for instance, while people are developing, they will not notice a problem; it means they will almost always have a round-trip failure through the buildfarm for this issue).

I'd rather we find another way to get the information we need into the setup scripts.  One thing I have done in the past is use subprocess() to call an external program to give me some additional information.  This ensures that there is no direct dependency in the setup.py script, but also allows us to get information from the system.  There may be other ways to go about this as well.

**Comment by wjwwood** [comment]
It might be possible to use `setup_requires` (http://setuptools.readthedocs.io/en/latest/setuptools.html#new-and-changed-setup-keywords) and move the imports and logic into steps that happen *after* `setup()` is called. It would require structural changes to how we modify the arguments to `setup()`, but it would also probably address our problem, since setuptools would download our dependencies and install them from `easy_install` (assuming we release `ament_python` on pypi.

If that is not sufficient (maybe the wrong version of `ament_python` is in PyPi), I think the only option is to actually install `ament_python` in the source deb environment.

**Comment by wjwwood** [comment]
Sorry, I should have said, I don't think gracefully falling back is a good idea, since in the case of installing binaries to `libexec`, the fallback behavior of not installing the binaries or installing them to `bin` is not graceful in my opinion, it is breaking the expected interface of the package.

**Comment by dirk-thomas** [comment]
I looked into using the `cmdclass` argument to the `setup` function. While it allows to provide a custom command it doesn't allow to use a class provided by a `setup_requires` dependency. Every `setup.py` file would need to define its own command class (or at least a class which dynamically redirect to an implementation defined in e.g. `ament_python`). Therefore I came to the conclusion that this is not feasible.

For now I will make PRs which replace the usage of `ament_python` with just performing the customization "inline" in each `setup.py` file. While that repeats some code it avoids the problem of requiring setup dependencies as well as make the code more readable to users (they will easily see what they get). See:
* ros2/demos#158
* ament/ament_tools#156

We could consider to propose changes to `setuptools` but at this time I am unsure how they could look like and therefore I haven't filled any ticket upstream yet.
